grpc/creds/creds.go

```diff
@@ -0,0 +1,88 @@
+package creds
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// transportCredentials is a grpc/credentials.TransportCredentials which supports
+// connecting to, and verifying multiple DNS names
+type transportCredentials struct {
+	configs map[string]*tls.Config
+}
+
+// New returns a new initialized grpc/credentials.TransportCredentials
+func New(addrs []string, rootCAs *x509.CertPool, clientCerts []tls.Certificate) (credentials.TransportCredentials, error) {
+	configs := make(map[string]*tls.Config, len(addrs))
+	for _, addr := range addrs {
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		configs[addr] = &tls.Config{
+			ServerName:   host,
+			RootCAs:      rootCAs,
+			Certificates: clientCerts,
+			MinVersion:   tls.VersionTLS12, // Override default of tls.VersionTLS10
+			MaxVersion:   tls.VersionTLS12, // Same as default in golang <= 1.6
+		}
+	}
+	return &transportCredentials{configs}, nil
+}
+
+// ClientHandshake performs the TLS handshake for a client -> server connection
+func (tc *transportCredentials) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	config, present := tc.configs[addr]
+	if !present {
+		return nil, nil, fmt.Errorf("boulder/grpc/creds: Unexpected name, no TLS configuration present for \"%s\"", addr)
+	}
+	conn := tls.Client(rawConn, config)
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- conn.Handshake()
+	}()
+	var after <-chan time.Time
```

Comment:
This seems a little less clear than the version I proposed:

```
if timeout == 0 {
  timeout = 1 * time.Hour
}
select {
  case <-time.After(timeout):
```

Mainly because the reader doesn't have to think about the fact that the default chan will block forever. Is there a reason you chose this particular implementation?